Can't install on Mac OS X 10.10

Version info and build log below:

```
$ uname -a
Darwin mbp.local 14.0.0 Darwin Kernel Version 14.0.0: Fri Sep 19 00:26:44 PDT 2014; root:xnu-2782.1.97~2/RELEASE_X86_64 x86_64

$ sudo python setup.py install
running install
running build
running build_py
creating build
creating build/lib.macosx-10.10-intel-2.7
creating build/lib.macosx-10.10-intel-2.7/dtrace_ctypes
copying dtrace_ctypes/__init__.py -> build/lib.macosx-10.10-intel-2.7/dtrace_ctypes
copying dtrace_ctypes/consumer.py -> build/lib.macosx-10.10-intel-2.7/dtrace_ctypes
copying dtrace_ctypes/dtrace_structs.py -> build/lib.macosx-10.10-intel-2.7/dtrace_ctypes
running build_ext
cythoning dtrace/dtrace.pyx to dtrace/dtrace.c

Error compiling Cython file:
------------------------------------------------------------
...
    def __init__(self, chew_func=None, chewrec_func=None, out_func=None,
                 walk_func=None):
        '''
        Constructor. Gets a DTrace handle and sets some options.
        '''
        self.chew_func = chew_func or simple_chew
                                                ^
------------------------------------------------------------

dtrace/dtrace.pyx:239:49: Cannot convert 'object (object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
    def __init__(self, chew_func=None, chewrec_func=None, out_func=None,
                 walk_func=None):
        '''
        Constructor. Gets a DTrace handle and sets some options.
        '''
        self.chew_func = chew_func or simple_chew
                                                ^
------------------------------------------------------------

dtrace/dtrace.pyx:239:49: Cannot convert 'object (object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
                 walk_func=None):
        '''
        Constructor. Gets a DTrace handle and sets some options.
        '''
        self.chew_func = chew_func or simple_chew
        self.chewrec_func = chewrec_func or simple_chewrec
                                                         ^
------------------------------------------------------------

dtrace/dtrace.pyx:240:58: Cannot convert 'object (object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
                 walk_func=None):
        '''
        Constructor. Gets a DTrace handle and sets some options.
        '''
        self.chew_func = chew_func or simple_chew
        self.chewrec_func = chewrec_func or simple_chewrec
                                                         ^
------------------------------------------------------------

dtrace/dtrace.pyx:240:58: Cannot convert 'object (object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
        '''
        Constructor. Gets a DTrace handle and sets some options.
        '''
        self.chew_func = chew_func or simple_chew
        self.chewrec_func = chewrec_func or simple_chewrec
        self.out_func = out_func or simple_out
                                             ^
------------------------------------------------------------

dtrace/dtrace.pyx:241:46: Cannot convert 'object (object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
        '''
        Constructor. Gets a DTrace handle and sets some options.
        '''
        self.chew_func = chew_func or simple_chew
        self.chewrec_func = chewrec_func or simple_chewrec
        self.out_func = out_func or simple_out
                                             ^
------------------------------------------------------------

dtrace/dtrace.pyx:241:46: Cannot convert 'object (object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
        Constructor. Gets a DTrace handle and sets some options.
        '''
        self.chew_func = chew_func or simple_chew
        self.chewrec_func = chewrec_func or simple_chewrec
        self.out_func = out_func or simple_out
        self.walk_func = walk_func or simple_walk
                                                ^
------------------------------------------------------------

dtrace/dtrace.pyx:242:49: Cannot convert 'object (object, object, object, object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
        Constructor. Gets a DTrace handle and sets some options.
        '''
        self.chew_func = chew_func or simple_chew
        self.chewrec_func = chewrec_func or simple_chewrec
        self.out_func = out_func or simple_out
        self.walk_func = walk_func or simple_walk
                                                ^
------------------------------------------------------------

dtrace/dtrace.pyx:242:49: Cannot convert 'object (object, object, object, object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
    def __init__(self, script, chew_func=None, chewrec_func=None,
                 out_func=None, walk_func=None):
        '''
        Constructor. will get the DTrace handle
        '''
        self.chew_func = chew_func or simple_chew
                                                ^
------------------------------------------------------------

dtrace/dtrace.pyx:354:49: Cannot convert 'object (object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
    def __init__(self, script, chew_func=None, chewrec_func=None,
                 out_func=None, walk_func=None):
        '''
        Constructor. will get the DTrace handle
        '''
        self.chew_func = chew_func or simple_chew
                                                ^
------------------------------------------------------------

dtrace/dtrace.pyx:354:49: Cannot convert 'object (object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
                 out_func=None, walk_func=None):
        '''
        Constructor. will get the DTrace handle
        '''
        self.chew_func = chew_func or simple_chew
        self.chewrec_func = chewrec_func or simple_chewrec
                                                         ^
------------------------------------------------------------

dtrace/dtrace.pyx:355:58: Cannot convert 'object (object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
                 out_func=None, walk_func=None):
        '''
        Constructor. will get the DTrace handle
        '''
        self.chew_func = chew_func or simple_chew
        self.chewrec_func = chewrec_func or simple_chewrec
                                                         ^
------------------------------------------------------------

dtrace/dtrace.pyx:355:58: Cannot convert 'object (object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
        '''
        Constructor. will get the DTrace handle
        '''
        self.chew_func = chew_func or simple_chew
        self.chewrec_func = chewrec_func or simple_chewrec
        self.out_func = out_func or simple_out
                                             ^
------------------------------------------------------------

dtrace/dtrace.pyx:356:46: Cannot convert 'object (object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
        '''
        Constructor. will get the DTrace handle
        '''
        self.chew_func = chew_func or simple_chew
        self.chewrec_func = chewrec_func or simple_chewrec
        self.out_func = out_func or simple_out
                                             ^
------------------------------------------------------------

dtrace/dtrace.pyx:356:46: Cannot convert 'object (object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
        Constructor. will get the DTrace handle
        '''
        self.chew_func = chew_func or simple_chew
        self.chewrec_func = chewrec_func or simple_chewrec
        self.out_func = out_func or simple_out
        self.walk_func = walk_func or simple_walk
                                                ^
------------------------------------------------------------

dtrace/dtrace.pyx:357:49: Cannot convert 'object (object, object, object, object, int __pyx_skip_dispatch)' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
        Constructor. will get the DTrace handle
        '''
        self.chew_func = chew_func or simple_chew
        self.chewrec_func = chewrec_func or simple_chewrec
        self.out_func = out_func or simple_out
        self.walk_func = walk_func or simple_walk
                                                ^
------------------------------------------------------------

dtrace/dtrace.pyx:357:49: Cannot convert 'object (object, object, object, object, int __pyx_skip_dispatch)' to Python object
building 'dtrace' extension
creating build/temp.macosx-10.10-intel-2.7
creating build/temp.macosx-10.10-intel-2.7/dtrace
cc -fno-strict-aliasing -fno-common -dynamic -arch x86_64 -arch i386 -g -Os -pipe -fno-common -fno-strict-aliasing -fwrapv -DENABLE_DTRACE -DMACOSX -DNDEBUG -Wall -Wstrict-prototypes -Wshorten-64-to-32 -DNDEBUG -g -fwrapv -Os -Wall -Wstrict-prototypes -DENABLE_DTRACE -arch x86_64 -arch i386 -pipe -I/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c dtrace/dtrace.c -o build/temp.macosx-10.10-intel-2.7/dtrace/dtrace.o
dtrace/dtrace.c:1:2: error: Do not use this file, it is the result of a failed Cython compilation.
#error Do not use this file, it is the result of a failed Cython compilation.
 ^
1 error generated.
error: command 'cc' failed with exit status 1
```

--- a/.gitignore
+++ b/.gitignore
@@ -11,3 +11,4 @@ out.txt
 dtrace/dtrace.c
 .tox
 .idea/
+/dtrace_cython/consumer.c

--- a/attic/dtrace.py
+++ b/attic/dtrace.py
@@ -4,6 +4,7 @@
 Simple Python based DTrace consumers which executes a Hello World D script
 using libdtrace and ctypes to access it.
 """
+from __future__ import print_function
 from ctypes import cdll, CDLL, Structure, c_void_p, c_char_p, c_uint, c_int, \
     c_char, CFUNCTYPE, POINTER, byref, cast
 import time
@@ -108,7 +109,7 @@ def chew_func(data, arg):
     """
     Callback for chew.
     """
-    print '+--> In chew: cpu :', c_int(data.contents.dtpda_cpu).value
+    print('+--> In chew: cpu :', c_int(data.contents.dtpda_cpu).value)
     return 0
 
 
@@ -126,7 +127,7 @@ def buffered_stdout_writer(bufdata, arg):
     In case dtrace_work is given None as filename - this one is called.
     """
     tmp = c_char_p(bufdata.contents.dtbda_buffered).value.strip()
-    print '  +--> In buffered_stdout_writer: ', tmp
+    print('  +--> In buffered_stdout_writer: ', tmp)
     return 0
 
 
@@ -141,7 +142,7 @@ def walk(data, arg):
     name = cast(tmp + 16, c_char_p).value
     instance = deref(tmp + 272, c_int).value
 
-    print '+--> walking', name, instance
+    print('+--> walking', name, instance)
 
     return 0
 
@@ -194,7 +195,7 @@ def run_dtrace():
 
     # Get errors if any...
     txt = LIBRARY.dtrace_errmsg(handle, LIBRARY.dtrace_errno(handle))
-    print c_char_p(txt).value
+    print(c_char_p(txt).value)
 
     # Last: close handle!
     LIBRARY.dtrace_close(handle)

--- a/dtrace_ctypes/consumer.py
+++ b/dtrace_ctypes/consumer.py
@@ -30,17 +30,17 @@ _LIBRARY = cdll.LoadLibrary(_LIBNAME)
 
 CHEW_FUNC = CFUNCTYPE(c_int,
                       POINTER(dtrace_probedata),
-                      POINTER(c_void_p))
+                      c_void_p)
 CHEWREC_FUNC = CFUNCTYPE(c_int,
                          POINTER(dtrace_probedata),
                          POINTER(dtrace_recdesc),
-                         POINTER(c_void_p))
+                         c_void_p)
 BUFFERED_FUNC = CFUNCTYPE(c_int,
                           POINTER(dtrace_bufdata),
-                          POINTER(c_void_p))
+                          c_void_p)
 WALK_FUNC = CFUNCTYPE(c_int,
                       POINTER(dtrace_aggdata),
-                      POINTER(c_void_p))
+                      c_void_p)
 
 
 def simple_chew_func(data, arg):

--- a/dtrace_ctypes/consumer.py
+++ b/dtrace_ctypes/consumer.py
@@ -11,12 +11,18 @@ from ctypes import cdll, CDLL, byref, c_int, c_char_p, CFUNCTYPE, c_void_p, \
 from dtrace_ctypes.dtrace_structs import dtrace_bufdata, dtrace_probedata, \
     dtrace_aggdata, dtrace_recdesc
 from threading import Thread
+import platform
 import threading
 import time
 
-cdll.LoadLibrary("libdtrace.so")
+if platform.system().startswith("Darwin"):
+    _LIBNAME = "libdtrace.dylib"
+else:
+    _LIBNAME = "libdtrace.so"
 
-LIBRARY = CDLL("libdtrace.so")
+cdll.LoadLibrary(_LIBNAME)
+
+LIBRARY = CDLL(_LIBNAME)
 
 # =============================================================================
 # chewing and output walkers

--- a/dtrace_ctypes/consumer.py
+++ b/dtrace_ctypes/consumer.py
@@ -15,7 +15,10 @@ from threading import Thread
 
 from dtrace_ctypes.dtrace_structs import (dtrace_aggdata, dtrace_bufdata,
                                           dtrace_hdl_t, dtrace_probedata,
-                                          dtrace_prog_t, dtrace_recdesc)
+                                          dtrace_prog_t, dtrace_recdesc,
+                                          DTRACE_WORKSTATUS_OKAY,
+                                          DTRACE_WORKSTATUS_DONE,
+                                          DTRACE_WORKSTATUS_ERROR)
 
 if platform.system().startswith("Darwin"):
     _LIBNAME = "libdtrace.dylib"
@@ -257,9 +260,14 @@ class DTraceConsumer(object):
         i = 0
         while i < runtime:
             LIBRARY.dtrace_sleep(self.handle)
-            LIBRARY.dtrace_work(self.handle, None, self.chew, self.chew_rec,
-                                None)
-
+            status = LIBRARY.dtrace_work(self.handle, None, self.chew,
+                                         self.chew_rec, None)
+            if status == DTRACE_WORKSTATUS_ERROR:
+                raise Exception('dtrace_work failed: ',
+                                get_error_msg(self.handle))
+            elif status == DTRACE_WORKSTATUS_DONE:
+                break  # No more work
+            assert status == DTRACE_WORKSTATUS_OKAY, status
             time.sleep(1)
             i += 1
 
@@ -349,9 +357,14 @@ class DTraceConsumerThread(Thread):
         # aggregate data for a few sec...
         while not self.stopped():
             LIBRARY.dtrace_sleep(self.handle)
-            LIBRARY.dtrace_work(self.handle, None, self.chew, self.chew_rec,
-                                None)
-
+            status = LIBRARY.dtrace_work(self.handle, None, self.chew,
+                                         self.chew_rec, None)
+            if status == DTRACE_WORKSTATUS_ERROR:
+                raise Exception('dtrace_work failed: ',
+                                get_error_msg(self.handle))
+            elif status == DTRACE_WORKSTATUS_DONE:
+                break  # No more work
+            assert status == DTRACE_WORKSTATUS_OKAY, status
             if LIBRARY.dtrace_aggregate_snap(self.handle) != 0:
                 raise Exception('Failed to snapshot the aggregate: ',
                                 get_error_msg(self.handle))

--- a/dtrace_ctypes/consumer.py
+++ b/dtrace_ctypes/consumer.py
@@ -5,24 +5,22 @@ Created on Oct 10, 2011
 
 @author: tmetsch
 """
-
-from ctypes import cdll, CDLL, byref, c_int, c_char_p, CFUNCTYPE, c_void_p, \
-    POINTER, cast
-from dtrace_ctypes.dtrace_structs import dtrace_bufdata, dtrace_probedata, \
-    dtrace_aggdata, dtrace_recdesc
-from threading import Thread
 import platform
 import threading
 import time
+from ctypes import (byref, c_char_p, c_int, c_uint, c_void_p, cast, cdll,
+                    CFUNCTYPE, POINTER)
+from threading import Thread
+
+from dtrace_ctypes.dtrace_structs import (dtrace_aggdata, dtrace_bufdata,
+                                          dtrace_hdl_t, dtrace_probedata,
+                                          dtrace_prog_t, dtrace_recdesc)
 
 if platform.system().startswith("Darwin"):
     _LIBNAME = "libdtrace.dylib"
 else:
     _LIBNAME = "libdtrace.so"
-
-cdll.LoadLibrary(_LIBNAME)
-
-LIBRARY = CDLL(_LIBNAME)
+_LIBRARY = cdll.LoadLibrary(_LIBNAME)
 
 # =============================================================================
 # chewing and output walkers
@@ -83,6 +81,61 @@ def simple_walk(data, arg):
     print '{0:60s} :{1:10d}'.format(name, instance)
 
     return 0
+
+# =============================================================================
+# LibDTrace function wrapper class
+# =============================================================================
+
+
+def _get_dtrace_fn(name, restype, argtypes):
+    fn = getattr(_LIBRARY, name)
+    fn.restype = restype
+    fn.argtypes = argtypes
+    return fn
+
+
+class LIBRARY(object):
+    # Types
+    dtrace_proginfo_t = c_void_p
+    dtrace_probespec_t = c_int  # actually an enum
+    dtrace_workstatus_t = c_int  # actually an enum
+    FILE_p = c_void_p  # FILE*
+    # Functions
+    dtrace_open = _get_dtrace_fn(
+        "dtrace_open", dtrace_hdl_t, [c_int, c_int, POINTER(c_int)])
+    dtrace_close = _get_dtrace_fn(
+        "dtrace_close", None, [dtrace_hdl_t])
+    dtrace_go = _get_dtrace_fn("dtrace_go", c_int, [dtrace_hdl_t])
+    dtrace_stop = _get_dtrace_fn("dtrace_stop", c_int, [dtrace_hdl_t])
+    dtrace_sleep = _get_dtrace_fn("dtrace_sleep", None, [dtrace_hdl_t])
+    dtrace_work = _get_dtrace_fn(
+        "dtrace_work", dtrace_workstatus_t,
+        [dtrace_hdl_t, FILE_p, CHEW_FUNC, CHEWREC_FUNC, c_void_p])
+    dtrace_errno = _get_dtrace_fn(
+        "dtrace_errno", c_int, [dtrace_hdl_t])
+    dtrace_errmsg = _get_dtrace_fn(
+        "dtrace_errmsg", c_char_p, [dtrace_hdl_t, c_int])
+    dtrace_setopt = _get_dtrace_fn(
+        "dtrace_setopt", c_int, [dtrace_hdl_t, c_char_p, c_char_p])
+    dtrace_handle_buffered = _get_dtrace_fn(
+        "dtrace_handle_buffered", c_int,
+        [dtrace_hdl_t, BUFFERED_FUNC, c_void_p])
+    dtrace_aggregate_walk = _get_dtrace_fn(
+        "dtrace_aggregate_walk", c_int,
+        [dtrace_hdl_t, WALK_FUNC, c_void_p])
+    dtrace_aggregate_walk_valsorted = _get_dtrace_fn(
+        "dtrace_aggregate_walk_valsorted", c_int,
+        [dtrace_hdl_t, WALK_FUNC, c_void_p])
+    dtrace_aggregate_snap = _get_dtrace_fn(
+        "dtrace_aggregate_snap", c_int, [dtrace_hdl_t])
+    dtrace_program_strcompile = _get_dtrace_fn(
+        "dtrace_program_strcompile", dtrace_prog_t,
+        [dtrace_hdl_t, c_char_p, dtrace_probespec_t, c_uint, c_int,
+         POINTER(c_char_p)])
+    dtrace_program_exec = _get_dtrace_fn(
+        "dtrace_program_exec", c_int,
+        [dtrace_hdl_t, dtrace_prog_t, dtrace_proginfo_t])
+
 
 # =============================================================================
 # Convenience stuff

--- a/dtrace_ctypes/consumer.py
+++ b/dtrace_ctypes/consumer.py
@@ -149,23 +149,39 @@ def deref(addr, typ):
     return cast(addr, POINTER(typ)).contents
 
 
-def get_error_msg(handle):
+def get_error_msg(handle, err=None):
     """
     Get the latest and greatest DTrace error.
     """
-    txt = LIBRARY.dtrace_errmsg(handle, LIBRARY.dtrace_errno(handle))
-    return c_char_p(txt).value
+    if err is None:
+        err = LIBRARY.dtrace_errno(handle)
+    txt = LIBRARY.dtrace_errmsg(handle, err)
+    return c_char_p(txt).value.decode("utf-8")
 
 # =============================================================================
 # Consumers
 # =============================================================================
+
+def _dtrace_open():
+    err = c_int(0)
+    handle = LIBRARY.dtrace_open(3, 0, byref(err))
+    if handle is None or handle == 0:
+        raise Exception('Unable to get a DTrace handle. Error=' +
+                        get_error_msg(handle, err))
+    assert isinstance(handle, dtrace_hdl_t)
+    # set buffer options
+    if LIBRARY.dtrace_setopt(handle, b'bufsize', b'4m') != 0:
+        raise Exception(get_error_msg(handle))
+
+    if LIBRARY.dtrace_setopt(handle, b'aggsize', b'4m') != 0:
+        raise Exception(get_error_msg(handle))
+    return handle
 
 
 class DTraceConsumer(object):
     """
     A Pyton based DTrace consumer.
     """
-
     def __init__(self,
                  chew_func=None,
                  chew_rec_func=None,
@@ -195,22 +211,14 @@ class DTraceConsumer(object):
             self.buf_out = BUFFERED_FUNC(simple_buffered_out_writer)
 
         # get dtrace handle
-        self.handle = LIBRARY.dtrace_open(3, 0, byref(c_int(0)))
-        if self.handle is None:
-            raise Exception('Unable to get a DTrace handle.')
-
-        # set buffer options
-        if LIBRARY.dtrace_setopt(self.handle, 'bufsize', '4m') != 0:
-            raise Exception(get_error_msg(self.handle))
-
-        if LIBRARY.dtrace_setopt(self.handle, 'aggsize', '4m') != 0:
-            raise Exception(get_error_msg(self.handle))
+        self.handle = _dtrace_open()
 
     def __del__(self):
         """
         Always close the DTrace handle :-)
         """
-        LIBRARY.dtrace_close(self.handle)
+        if hasattr(self, "handle"):  # Don't close if __init__ failed.
+            LIBRARY.dtrace_close(self.handle)
 
     def run(self, script, runtime=1):
         """
@@ -224,6 +232,7 @@ class DTraceConsumer(object):
         runtime -- The time the script should run in second (Default: 1s).
         """
         # set simple output callbacks
+        assert self.handle is not None
         if LIBRARY.dtrace_handle_buffered(self.handle, self.buf_out,
                                           None) == -1:
             raise Exception('Unable to set the stdout buffered writer.')
@@ -304,25 +313,18 @@ class DTraceConsumerThread(Thread):
             self.buf_out = BUFFERED_FUNC(simple_buffered_out_writer)
 
         # get dtrace handle
-        self.handle = LIBRARY.dtrace_open(3, 0, byref(c_int(0)))
-        if self.handle is None:
-            raise Exception('Unable to get a DTrace handle.')
-
-        # set buffer options
-        if LIBRARY.dtrace_setopt(self.handle, 'bufsize', '4m') != 0:
-            raise Exception(get_error_msg(self.handle))
-
-        if LIBRARY.dtrace_setopt(self.handle, 'aggsize', '4m') != 0:
-            raise Exception(get_error_msg(self.handle))
+        self.handle = _dtrace_open()
 
     def __del__(self):
         """
         Always close the DTrace handle :-)
         """
-        LIBRARY.dtrace_close(self.handle)
+        if hasattr(self, "handle"):  # Don't close if __init__ failed.
+            LIBRARY.dtrace_close(self.handle)
 
     def run(self):
         Thread.run(self)
+        assert self.handle is not None
         # set simple output callbacks
         if LIBRARY.dtrace_handle_buffered(self.handle, self.buf_out,
                                           None) == -1:

--- a/dtrace_ctypes/consumer.py
+++ b/dtrace_ctypes/consumer.py
@@ -238,8 +238,8 @@ class DTraceConsumer(object):
             raise Exception('Unable to set the stdout buffered writer.')
 
         # compile
-        prg = LIBRARY.dtrace_program_strcompile(self.handle,
-                                                script, 3, 4, 0, None)
+        prg = LIBRARY.dtrace_program_strcompile(
+            self.handle, c_char_p(script.encode("utf-8")), 3, 4, 0, None)
         if prg is None:
             raise Exception('Unable to compile the script: ',
                             get_error_msg(self.handle))
@@ -266,8 +266,7 @@ class DTraceConsumer(object):
 
         # sorting instead of dtrace_aggregate_walk
 
-        if LIBRARY.dtrace_aggregate_walk_valsorted(self.handle,
-                                                   self.walk,
+        if LIBRARY.dtrace_aggregate_walk_valsorted(self.handle, self.walk,
                                                    None) != 0:
             raise Exception('Failed to walk the aggregate: ',
                             get_error_msg(self.handle))
@@ -331,9 +330,9 @@ class DTraceConsumerThread(Thread):
             raise Exception('Unable to set the stdout buffered writer.')
 
         # compile
-        prg = LIBRARY.dtrace_program_strcompile(self.handle,
-                                                self.script, 3, 4, 0,
-                                                None)
+        prg = LIBRARY.dtrace_program_strcompile(
+            self.handle, c_char_p(self.script.encode("utf-8")), 3, 4, 0,
+            None)
         if prg is None:
             raise Exception('Unable to compile the script: ',
                             get_error_msg(self.handle))

--- a/dtrace_ctypes/consumer.py
+++ b/dtrace_ctypes/consumer.py
@@ -5,6 +5,7 @@ Created on Oct 10, 2011
 
 @author: tmetsch
 """
+from __future__ import print_function
 import platform
 import threading
 import time
@@ -78,7 +79,7 @@ def simple_walk(data, arg):
     name = cast(tmp + 16, c_char_p).value
     instance = deref(tmp + 272, c_int).value
 
-    print '{0:60s} :{1:10d}'.format(name, instance)
+    print('{0:60s} :{1:10d}'.format(name, instance))
 
     return 0
 

--- a/dtrace_ctypes/dtrace_structs.py
+++ b/dtrace_ctypes/dtrace_structs.py
@@ -10,8 +10,18 @@ Created on Oct 10, 2011
 # disabling 'Invalid name' pylint check (C0103)
 # pylint: disable=R0903,C0103
 
-from ctypes import Structure, c_int, c_void_p, c_char_p, c_char, c_uint
+from ctypes import Structure, c_int, c_void_p, c_char_p, c_char, c_uint, POINTER
 
+class dtrace_hdl(Structure):
+    pass
+
+
+class dtrace_prog(Structure):
+    pass
+
+
+dtrace_hdl_t = POINTER(dtrace_hdl)
+dtrace_prog_t = POINTER(dtrace_prog)
 
 class dtrace_recdesc(Structure):
     """

--- a/dtrace_ctypes/dtrace_structs.py
+++ b/dtrace_ctypes/dtrace_structs.py
@@ -10,7 +10,9 @@ Created on Oct 10, 2011
 # disabling 'Invalid name' pylint check (C0103)
 # pylint: disable=R0903,C0103
 
-from ctypes import Structure, c_int, c_void_p, c_char_p, c_char, c_uint, POINTER
+from ctypes import (c_int64, c_size_t, c_uint16, c_uint32, c_uint64, POINTER,
+                    Structure, c_int, c_void_p, c_char_p, c_char, c_uint)
+
 
 class dtrace_hdl(Structure):
     pass
@@ -27,46 +29,30 @@ class dtrace_recdesc(Structure):
     """
     sys/dtrace.h:931
     """
-    _fields_ = [("dtrd_offset", c_uint)]
+    _fields_ = [
+        ("dtrd_action", c_uint16),  # dtrace_actkind_t
+        ("dtrd_size", c_uint32),
+        ("dtrd_offset", c_uint32),
+        ("dtrd_alignment", c_uint16),
+        ("dtrd_format", c_uint16),
+        ("dtrd_arg", c_uint64),
+        ("dtrd_uarg", c_uint64)]
 
 
 class dtrace_aggdesc(Structure):
     """
     sys/dtrace.h:950
     """
-    _fields_ = [("dtagd_rec", c_void_p)]
-
-
-class dtrace_aggdata(Structure):
-    """
-    As defined in dtrace.h:351
-    """
-    _fields_ = [("dtada_handle", c_void_p),
-                ("dtada_desc", dtrace_aggdesc),
-                ("dtada_edesc", c_void_p),
-                ("dtada_pdesc", c_void_p),
-                ("dtada_data", c_void_p),
-                ("dtada_normal", c_void_p),
-                ("dtada_size", c_int),
-                ("dtada_delta", c_void_p),
-                ("dtada_percpu", c_void_p),
-                ("dtada_percpu_delta", c_void_p),
-                ("dtada_total", c_void_p),
-                ("dtada_minbin", c_void_p),
-                ("dtada_maxbin", c_void_p),
-                ("dtada_flags", c_void_p)]
-
-
-class dtrace_bufdata(Structure):
-    """
-    As defined in dtrace.h:310
-    """
-    _fields_ = [("dtbda_handle", c_void_p),
-                ("dtbda_buffered", c_char_p),
-                ("dtbda_probe", c_void_p),
-                ("dtbda_recdesc", c_void_p),
-                ("dtbda_aggdata", c_void_p),
-                ("dtbda_flags", c_uint)]
+    _fields_ = [("dtagd_name", c_char_p),
+                ("dtagd_varid", c_int64),  # dtrace_aggvarid_t
+                ("dtagd_flags", c_int),
+                ("dtagd_id", c_void_p),  # dtrace_aggid_t
+                ("dtagd_epid", c_void_p),  # dtrace_epid_t
+                ("dtagd_size", c_uint32),
+                ("dtagd_nrecs", c_int),
+                ("dtagd_pad", c_uint32),
+                ("dtagd_rec", dtrace_recdesc),  # variable-length
+                ]
 
 
 class dtrace_probedesc(Structure):
@@ -80,18 +66,51 @@ class dtrace_probedesc(Structure):
                 ("dtpd_name", c_char_p)]
 
 
+class dtrace_aggdata(Structure):
+    """
+    As defined in dtrace.h:351
+    """
+    _fields_ = [("dtada_handle", dtrace_hdl_t),
+                ("dtada_desc", POINTER(dtrace_aggdesc)),
+                ("dtada_edesc", c_void_p),
+                ("dtada_pdesc", POINTER(dtrace_probedesc)),
+                ("dtada_data", c_void_p),
+                ("dtada_normal", c_uint64),
+                ("dtada_size", c_size_t),
+                ("dtada_delta", c_void_p),
+                ("dtada_percpu", c_void_p),
+                ("dtada_percpu_delta", c_void_p),
+                ("dtada_total", c_int64),
+                ("dtada_minbin", c_uint16),
+                ("dtada_maxbin", c_uint16),
+                ("dtada_flags", c_uint32)]
+
+
 class dtrace_probedata(Structure):
     """
     As defined in dtrace.h:186
     """
-    _fields_ = [("dtpda_handle", c_void_p),
+    _fields_ = [("dtpda_handle", dtrace_hdl_t),
                 ("dtpda_edesc", c_void_p),
-                ("dtpda_pdesc", dtrace_probedesc),
+                ("dtpda_pdesc", POINTER(dtrace_probedesc)),
                 ("dtpda_cpu", c_int),
                 ("dtpda_data", c_void_p),
-                ("dtpda_flow", c_void_p),
-                ("dtpda_prefix", c_void_p),
+                ("dtpda_flow", c_int),  # dtrace_flowkind_t
+                ("dtpda_prefix", c_char_p),
                 ("dtpda_indent", c_int)]
+
+
+class dtrace_bufdata(Structure):
+    """
+    As defined in dtrace.h:310
+    """
+    _fields_ = [("dtbda_handle", dtrace_hdl_t),
+                ("dtbda_buffered", c_char_p),
+                ("dtbda_probe", POINTER(dtrace_probedata)),
+                ("dtbda_recdesc", POINTER(dtrace_recdesc)),
+                ("dtbda_aggdata", POINTER(dtrace_aggdata)),
+                ("dtbda_flags", c_uint)]
+
 
 DTRACE_WORKSTATUS_ERROR = -1
 DTRACE_WORKSTATUS_OKAY = 0

--- a/dtrace_ctypes/dtrace_structs.py
+++ b/dtrace_ctypes/dtrace_structs.py
@@ -92,3 +92,7 @@ class dtrace_probedata(Structure):
                 ("dtpda_flow", c_void_p),
                 ("dtpda_prefix", c_void_p),
                 ("dtpda_indent", c_int)]
+
+DTRACE_WORKSTATUS_ERROR = -1
+DTRACE_WORKSTATUS_OKAY = 0
+DTRACE_WORKSTATUS_DONE = 1

--- a/dtrace_cython/consumer.pyx
+++ b/dtrace_cython/consumer.pyx
@@ -265,21 +265,21 @@ cdef class DTraceConsumer:
         if self.handle != NULL:
             dtrace_close(self.handle)
 
-    cpdef compile(self, char * script):
+    cpdef compile(self, unicode script):
         """
         Compile a DTrace script and return errors if any.
         
         script -- The script to compile.
         """
         cdef dtrace_prog_t * prg
-        prg = dtrace_program_strcompile(self.handle, script,
+        prg = dtrace_program_strcompile(self.handle, script.encode("utf-8"),
                                         DTRACE_PROBESPEC_NAME, 0, 0, NULL)
         if prg == NULL:
             raise Exception('Unable to compile the script: ',
                             dtrace_errmsg(self.handle,
                                           dtrace_errno(self.handle)))
 
-    cpdef run(self, char * script, runtime=1):
+    cpdef run(self, unicode script, runtime=1):
         """
         Run a DTrace script for a number of seconds defined by the runtime.
 
@@ -297,7 +297,7 @@ cdef class DTraceConsumer:
 
         # compile
         cdef dtrace_prog_t * prg
-        prg = dtrace_program_strcompile(self.handle, script,
+        prg = dtrace_program_strcompile(self.handle, script.encode("utf-8"),
                                         DTRACE_PROBESPEC_NAME, 0, 0, NULL)
         if prg == NULL:
             raise Exception('Unable to compile the script: ',
@@ -355,7 +355,7 @@ cdef class DTraceContinuousConsumer:
     cdef object chewrec_func
     cdef object script
 
-    def __init__(self, script, chew_func=None, chewrec_func=None,
+    def __init__(self, unicode script, chew_func=None, chewrec_func=None,
                  out_func=None, walk_func=None):
         """
         Constructor. will get the DTrace handle
@@ -364,7 +364,7 @@ cdef class DTraceContinuousConsumer:
         self.chewrec_func = chewrec_func or simple_chewrec
         self.out_func = out_func or simple_out
         self.walk_func = walk_func or simple_walk
-        self.script = script
+        self.script = script.encode("utf-8")
 
         cdef int err
         if not hasattr(self, 'handle'):

--- a/dtrace_cython/consumer.pyx
+++ b/dtrace_cython/consumer.pyx
@@ -3,6 +3,7 @@ import time
 import threading
 from threading import Thread
 from dtrace_cython.dtrace_h cimport *
+from libc.stdint cimport INT64_MAX, INT64_MIN
 
 # ----------------------------------------------------------------------------
 # The DTrace callbacks

--- a/dtrace_cython/consumer.pyx
+++ b/dtrace_cython/consumer.pyx
@@ -9,7 +9,7 @@ from dtrace_cython.dtrace_h cimport *
 # ----------------------------------------------------------------------------
 
 
-cdef int chew(dtrace_probedata_t * data, void * arg) with gil:
+cdef int chew(const dtrace_probedata_t * data, void * arg) with gil:
     """
     Callback defined by DTrace - will call the Python callback.
 
@@ -26,7 +26,7 @@ cdef int chew(dtrace_probedata_t * data, void * arg) with gil:
     return 0
 
 
-cdef int chewrec(dtrace_probedata_t * data, dtrace_recdesc_t * rec,
+cdef int chewrec(const dtrace_probedata_t * data, const dtrace_recdesc_t * rec,
                  void * arg) with gil:
     """
     Callback defined by DTrace - will call the Python callback.
@@ -46,7 +46,7 @@ cdef int chewrec(dtrace_probedata_t * data, dtrace_recdesc_t * rec,
     return 0
 
 
-cdef int buf_out(dtrace_bufdata_t * buf_data, void * arg) with gil:
+cdef int buf_out(const dtrace_bufdata_t * buf_data, void * arg) with gil:
     """
     Callback defined by DTrace - will call the Python callback.
     """
@@ -59,7 +59,7 @@ cdef int buf_out(dtrace_bufdata_t * buf_data, void * arg) with gil:
     return 0
 
 
-cdef int walk(dtrace_aggdata_t * data, void * arg) with gil:
+cdef int walk(const dtrace_aggdata_t * data, void * arg) with gil:
     """
     Callback defined by DTrace - will call the Python callback.
     """

--- a/dtrace_cython/consumer.pyx
+++ b/dtrace_cython/consumer.pyx
@@ -1,4 +1,4 @@
-
+from __future__ import print_function, division
 import time
 import threading
 from threading import Thread
@@ -90,7 +90,7 @@ cdef int walk(const dtrace_aggdata_t * data, void * arg) with gil:
         value = (<int *>(data.dtada_data + aggrec.dtrd_offset))[0]
     elif action == DTRACEAGG_AVG:
         tmp = <int64_t *>(data.dtada_data + aggrec.dtrd_offset)
-        value = tmp[1] / tmp[0]
+        value = tmp[1] // tmp[0]
     elif action == DTRACEAGG_QUANTIZE:
         tmp = <int64_t *>(data.dtada_data + aggrec.dtrd_offset)
         ranges = get_quantize_ranges()
@@ -105,7 +105,7 @@ cdef int walk(const dtrace_aggdata_t * data, void * arg) with gil:
         tmp_arg = tmp[0]
 
         ranges = get_lquantize_ranges(tmp_arg)
-        levels = (aggrec.dtrd_size / sizeof (uint64_t)) - 1
+        levels = (aggrec.dtrd_size // sizeof (uint64_t)) - 1
         for i in range(0, levels):
             # i + 1 since tmp[0] is already 'used'
             quantize.append((ranges[i], tmp[i + 1]))
@@ -182,7 +182,7 @@ def simple_chew(cpu):
 
     cpu -- CPU id.
     """
-    print 'Running on CPU:', cpu
+    print('Running on CPU:', cpu)
 
 
 def simple_chewrec(action):
@@ -191,7 +191,7 @@ def simple_chewrec(action):
 
     action -- id of the action which was called.
     """
-    print 'Called action was:', action
+    print('Called action was:', action)
 
 
 def simple_out(value):
@@ -200,7 +200,7 @@ def simple_out(value):
 
     value -- Line by line string of the DTrace output.
     """
-    print 'Value is:', value
+    print('Value is:', value)
 
 
 def simple_walk(action, identifier, keys, value):
@@ -213,7 +213,7 @@ def simple_walk(action, identifier, keys, value):
     keys -- list of keys.
     value -- the value.
     """
-    print action, identifier, keys, value
+    print(action, identifier, keys, value)
 
 # ----------------------------------------------------------------------------
 # The consumers

--- a/dtrace_cython/consumer.pyx
+++ b/dtrace_cython/consumer.pyx
@@ -321,9 +321,14 @@ cdef class DTraceConsumer:
             dtrace_sleep(self.handle)
             status = dtrace_work(self.handle, NULL, & chew, & chewrec,
                                  <void *>args)
-            if status == 1:
-                i = runtime
+            if status == DTRACE_WORKSTATUS_DONE:
+                break
+            elif status == DTRACE_WORKSTATUS_ERROR:
+                raise Exception('dtrace_work failed: ',
+                                dtrace_errmsg(self.handle,
+                                              dtrace_errno(self.handle)))
             else:
+                assert status == DTRACE_WORKSTATUS_OKAY, status
                 time.sleep(1)
                 i += 1
                 
@@ -428,7 +433,10 @@ cdef class DTraceContinuousConsumer:
         args = (self.chew_func, self.chewrec_func)
         status = dtrace_work(self.handle, NULL, & chew, & chewrec,
                              <void *>args)
-
+        if status == DTRACE_WORKSTATUS_ERROR:
+            raise Exception('dtrace_work failed: ',
+                            dtrace_errmsg(self.handle,
+                                          dtrace_errno(self.handle)))
         if dtrace_aggregate_snap(self.handle) != 0:
             raise Exception('Failed to get the aggregate: ',
                             dtrace_errmsg(self.handle,

--- a/dtrace_cython/consumer.pyx
+++ b/dtrace_cython/consumer.pyx
@@ -154,9 +154,9 @@ cdef get_quantize_ranges():
 cdef get_lquantize_ranges(uint64_t arg):
     ranges = []
 
-    base = DTRACE_LQUANTIZE_BASE(arg);
-    step = DTRACE_LQUANTIZE_STEP(arg);
-    levels = DTRACE_LQUANTIZE_LEVELS(arg);
+    base = DTRACE_LQUANTIZE_BASE(arg)
+    step = DTRACE_LQUANTIZE_STEP(arg)
+    levels = DTRACE_LQUANTIZE_LEVELS(arg)
 
     for i in range(0, levels + 2):
         if i == 0:

--- a/dtrace_cython/dtrace_h.pxd
+++ b/dtrace_cython/dtrace_h.pxd
@@ -35,9 +35,9 @@ cdef extern from "sys/dtrace.h":
 
     ctypedef struct dtrace_recdesc_t:
         # Taken from sys/dtrace.h:931
-        int dtrd_action
-        int dtrd_offset
-        int dtrd_size
+        uint16_t dtrd_action
+        uint32_t dtrd_offset
+        uint32_t dtrd_size
 
     ctypedef struct dtrace_aggdesc_t:
         # Taken from sys/dtrace.h:950
@@ -45,10 +45,10 @@ cdef extern from "sys/dtrace.h":
         int64_t dtagd_varid
         dtrace_recdesc_t dtagd_rec[1]
 
-    cdef uint16_t DTRACE_LQUANTIZE_STEP(long x)
-    cdef uint16_t DTRACE_LQUANTIZE_LEVELS(long x)
-    cdef int32_t DTRACE_LQUANTIZE_BASE(long x)
-    cdef int64_t DTRACE_QUANTIZE_BUCKETVAL(long buck)
+    cdef uint16_t DTRACE_LQUANTIZE_STEP(uint64_t x)
+    cdef uint16_t DTRACE_LQUANTIZE_LEVELS(uint64_t x)
+    cdef int32_t DTRACE_LQUANTIZE_BASE(uint64_t x)
+    cdef int64_t DTRACE_QUANTIZE_BUCKETVAL(uint64_t buck)
 
 
 cdef extern from "dtrace.h":

--- a/dtrace_cython/dtrace_h.pxd
+++ b/dtrace_cython/dtrace_h.pxd
@@ -48,7 +48,7 @@ cdef extern from "sys/dtrace.h":
     ctypedef struct dtrace_aggdesc_t:
         # Taken from sys/dtrace.h:950
         int dtagd_nrecs
-        int dtagd_varid
+        int64_t dtagd_varid
         dtrace_recdesc_t dtagd_rec[1]
 
     cdef uint16_t DTRACE_LQUANTIZE_STEP(long x)

--- a/dtrace_cython/dtrace_h.pxd
+++ b/dtrace_cython/dtrace_h.pxd
@@ -20,24 +20,18 @@ ELSE:
 
 
 cdef extern from "sys/dtrace.h":
+    uint64_t DTRACE_QUANTIZE_NBUCKETS
+    uint64_t DTRACE_QUANTIZE_ZEROBUCKET
 
-    ctypedef enum agg_actions:
-        # Taken from sys/dtrace.h:454
-        # Needs to be in enum because ctypes wants it that way :-/
-        DTRACEACT_AGGREGATION = 0x0700
-        DTRACEAGG_COUNT = (DTRACEACT_AGGREGATION + 1)
-        DTRACEAGG_MIN = (DTRACEACT_AGGREGATION + 2)
-        DTRACEAGG_MAX = (DTRACEACT_AGGREGATION + 3)
-        DTRACEAGG_AVG = (DTRACEACT_AGGREGATION + 4)
-        DTRACEAGG_SUM = (DTRACEACT_AGGREGATION + 5)
-        # + 6 is DTRACEAGG_STDDEV (unsupported)
-        DTRACEAGG_QUANTIZE = (DTRACEACT_AGGREGATION + 7)
-        DTRACEAGG_LQUANTIZE = (DTRACEACT_AGGREGATION + 8)
-
-    ctypedef enum quantize_types:
-        # NBBY = 8
-        DTRACE_QUANTIZE_NBUCKETS = (((sizeof (uint64_t) * 8) - 1) * 2 + 1)
-        DTRACE_QUANTIZE_ZEROBUCKET = ((sizeof (uint64_t) * 8) - 1)
+    int DTRACEACT_AGGREGATION
+    int DTRACEAGG_COUNT
+    int DTRACEAGG_MIN
+    int DTRACEAGG_MAX
+    int DTRACEAGG_AVG
+    int DTRACEAGG_SUM
+    int DTRACEAGG_STDDEV # (unsupported)
+    int DTRACEAGG_QUANTIZE
+    int DTRACEAGG_LQUANTIZE
 
     ctypedef struct dtrace_recdesc_t:
         # Taken from sys/dtrace.h:931
@@ -61,16 +55,16 @@ cdef extern from "dtrace.h":
 
     ctypedef enum dtrace_probespec_t:
         # Taken from dtrace.h:186
-        DTRACE_PROBESPEC_NONE = -1
-        DTRACE_PROBESPEC_PROVIDER = 0
+        DTRACE_PROBESPEC_NONE
+        DTRACE_PROBESPEC_PROVIDER
         DTRACE_PROBESPEC_MOD
         DTRACE_PROBESPEC_FUNC
         DTRACE_PROBESPEC_NAME
 
     ctypedef enum dtrace_workstatus_t:
         # Taken from dtrace.h:247
-        DTRACE_WORKSTATUS_ERROR = -1,
-        DTRACE_WORKSTATUS_OKAY,
+        DTRACE_WORKSTATUS_ERROR
+        DTRACE_WORKSTATUS_OKAY
         DTRACE_WORKSTATUS_DONE
 
     ctypedef struct dtrace_hdl_t:

--- a/dtrace_cython/dtrace_h.pxd
+++ b/dtrace_cython/dtrace_h.pxd
@@ -7,18 +7,6 @@ cdef extern from "libelf_workaround.h":
     pass
 
 
-IF UNAME_SYSNAME == "Darwin":
-    cdef extern from "stdint.h":
-        # needed for quantize mths.
-        cdef int64_t INT64_MAX
-        cdef int64_t INT64_MIN
-ELSE:
-    cdef extern from "sys/int_limits.h":
-        # needed for quantize mths.
-        cdef int64_t INT64_MAX
-        cdef int64_t INT64_MIN
-
-
 cdef extern from "sys/dtrace.h":
     uint64_t DTRACE_QUANTIZE_NBUCKETS
     uint64_t DTRACE_QUANTIZE_ZEROBUCKET

--- a/dtrace_cython/dtrace_h.pxd
+++ b/dtrace_cython/dtrace_h.pxd
@@ -79,13 +79,24 @@ cdef extern from "dtrace.h":
         # Taken from dtrace.h:97
         pass
 
+    ctypedef struct dtrace_probedesc_t:
+        uint32_t id
+        char dtpd_provider[64]  # DTRACE_PROVNAMELEN
+        char dtpd_mod[64]  # DTRACE_MODNAMELEN
+        char dtpd_func[192]  # DTRACE_FUNCNAMELEN
+        char dtpd_name[64]  # DTRACE_NAMELEN
+
     ctypedef struct dtrace_probedata_t:
         # Taken from dtrace.h:186
+        dtrace_hdl_t * dtbda_handle
+        dtrace_probedesc_t * dtpda_pdesc
         int dtpda_cpu
 
     ctypedef struct dtrace_bufdata_t:
         # Taken from dtrace.h:310
-        char * dtbda_buffered
+        dtrace_hdl_t * dtbda_handle
+        const char * dtbda_buffered
+        dtrace_probedata_t * dtbda_probe
 
     ctypedef struct dtrace_aggdata_t:
         # Taken from dtrace.h:351

--- a/dtrace_cython/dtrace_h.pxd
+++ b/dtrace_cython/dtrace_h.pxd
@@ -99,10 +99,10 @@ cdef extern from "dtrace.h":
         char * dtada_data
 
     # from dtrace.h
-    ctypedef int dtrace_handle_buffered_f(dtrace_bufdata_t * buf_data, void * arg)
-    ctypedef int dtrace_consume_probe_f(dtrace_probedata_t * , void *)
-    ctypedef int dtrace_consume_rec_f(dtrace_probedata_t * , dtrace_recdesc_t * , void *)
-    ctypedef int dtrace_aggregate_f(dtrace_aggdata_t * , void *)
+    ctypedef int dtrace_handle_buffered_f(const dtrace_bufdata_t * buf_data, void * arg)
+    ctypedef int dtrace_consume_probe_f(const dtrace_probedata_t *, void *)
+    ctypedef int dtrace_consume_rec_f(const dtrace_probedata_t *, const dtrace_recdesc_t * , void *)
+    ctypedef int dtrace_aggregate_f(const dtrace_aggdata_t *, void *)
 
     # open and close handle
     dtrace_hdl_t * dtrace_open(int, int, int *)
@@ -113,9 +113,9 @@ cdef extern from "dtrace.h":
 
     # output handling
     int dtrace_handle_buffered(dtrace_hdl_t * , dtrace_handle_buffered_f * , void *)
-    int dtrace_consume_probe_f(dtrace_probedata_t * , void *)
-    int dtrace_consume_rec_f(dtrace_probedata_t * , dtrace_recdesc_t * , void *)
-    int dtrace_aggregate_f(dtrace_aggdata_t * , void *)
+    int dtrace_consume_probe_f(const dtrace_probedata_t * , void *)
+    int dtrace_consume_rec_f(const dtrace_probedata_t * , dtrace_recdesc_t * , void *)
+    int dtrace_aggregate_f(const dtrace_aggdata_t * , void *)
 
     # compile
     dtrace_prog_t * dtrace_program_strcompile(dtrace_hdl_t * , char * , dtrace_probespec_t, int, int, char * const [])

--- a/examples/amqp/receive.py
+++ b/examples/amqp/receive.py
@@ -7,7 +7,7 @@ Created on Mar 31, 2012
 
 @author: tmetsch
 """
-
+from __future__ import print_function
 import ast
 import pika
 
@@ -34,7 +34,7 @@ def callback(ch, method, properties, body):
         data[key] = val
     # Print instead of doing print you could life update a chart...msg will
     # come as DTrace fires them
-    print 'Received: ', key, val
+    print('Received: ', key, val)
 
 if __name__ == '__main__':
     channel.basic_consume(callback,

--- a/examples/amqp/receive.py
+++ b/examples/amqp/receive.py
@@ -37,9 +37,9 @@ def callback(ch, method, properties, body):
     print('Received: ', key, val)
 
 if __name__ == '__main__':
-    channel.basic_consume(callback,
+    channel.basic_consume(on_message_callback=callback,
                           queue='dtrace',
-                          no_ack=True)
+                          auto_ack=False)
 
     try:
         channel.start_consuming()

--- a/examples/amqp/send.py
+++ b/examples/amqp/send.py
@@ -28,7 +28,7 @@ def walk(iden, key, value):
     Walker which sends on the data to RabbitMQ as the DTrace probes fire.
     """
     channel.basic_publish(exchange='', routing_key='dtrace',
-                          body=str({key[0]: value}))
+                          body=bytes({key[0]: value}))
 
 
 def run_dtrace():

--- a/examples/cli.py
+++ b/examples/cli.py
@@ -8,7 +8,7 @@ Created on Apr 11, 2012
 
 @author: tmetsch
 """
-
+from __future__ import print_function
 from dtrace import DTraceConsumerThread
 import sys
 
@@ -25,12 +25,12 @@ def print_lquantize(values):
 
     for item in values:
         if item[0][0] > 0:
-            print '%10s ' % item[0][0],
+            print('%10s ' % item[0][0], end=' ')
             for i in range(0, ((40 * int(item[1])) / maxi)):
                 sys.stdout.write('*')
             for i in range(((40 * int(item[1])) / maxi), 40):
                 sys.stdout.write(' ')
-            print ' %5s' % item[1]
+            print(' %5s' % item[1])
 
 
 def pretty_print(iden, action, keys, values):
@@ -38,12 +38,12 @@ def pretty_print(iden, action, keys, values):
     Pretty print aggregation walker.
     """
     if action in [1799]:
-        print keys, values
+        print(keys, values)
     elif action == 1800:
         # lquantize
-        print '\n    ', keys[0], '\n'
-        print '{0:>10s} {1:-^40} {2}'.format('value', ' Distribution ',
-                                             'count')
+        print('\n    ', keys[0], '\n')
+        print('{0:>10s} {1:-^40} {2}'.format('value', ' Distribution ',
+                                             'count'))
         print_lquantize(values)
     else:
         pass
@@ -53,7 +53,7 @@ def brendan():
     """
     DTrace fans will understand this :-D
     """
-    print 'Tracing... Hit Ctrl-C to end'
+    print('Tracing... Hit Ctrl-C to end')
 
 
 def run_dtrace(script):

--- a/examples/ctypes/read_bytes.py
+++ b/examples/ctypes/read_bytes.py
@@ -7,7 +7,7 @@ Created on Oct 10, 2011
 
 @author: tmetsch
 """
-
+from __future__ import print_function
 from dtrace_ctypes.consumer import DTraceConsumer
 
 SCRIPT = 'sysinfo:::readch { @bytes[execname] = sum(arg0); }'
@@ -17,7 +17,7 @@ def main():
     """
     Run DTrace...
     """
-    print 'Hint: if you don\'t get any output try running it with pfexec...'
+    print('Hint: if you don\'t get any output try running it with pfexec...')
 
     consumer = DTraceConsumer()
     consumer.run(SCRIPT, 4)

--- a/examples/ctypes/syscall_by_zone.py
+++ b/examples/ctypes/syscall_by_zone.py
@@ -8,7 +8,7 @@ Created on Oct 10, 2011
 @author: tmetsch
 """
 
-
+from __future__ import print_function
 from ctypes import cast, c_char_p, c_int
 from dtrace_ctypes.consumer import DTraceConsumerThread, deref
 import time
@@ -25,7 +25,7 @@ def walk(data, arg):
     name = cast(tmp + 16, c_char_p).value
     count = deref(tmp + 272, c_int).value
 
-    print 'Zone "{0:s}" made {1:d} syscalls.'.format(name, count)
+    print('Zone "{0:s}" made {1:d} syscalls.'.format(name, count))
 
     return 0
 

--- a/examples/ctypes/syscall_count_own_walk.py
+++ b/examples/ctypes/syscall_count_own_walk.py
@@ -8,7 +8,7 @@ Created on Oct 10, 2011
 
 @author: tmetsch
 """
-
+from __future__ import print_function
 from ctypes import cast, c_char_p, c_int
 from dtrace_ctypes.consumer import DTraceConsumer, deref
 
@@ -25,7 +25,7 @@ def my_walk(data, arg):
     name = cast(tmp + 16, c_char_p).value
     instance = deref(tmp + 272, c_int).value
 
-    print '{0:4d} > {1:60s}'.format(instance, name)
+    print('{0:4d} > {1:60s}'.format(instance, name))
 
     return 0
 

--- a/examples/read_bytes.py
+++ b/examples/read_bytes.py
@@ -7,7 +7,7 @@ Created on Oct 10, 2011
 
 @author: tmetsch
 """
-
+from __future__ import print_function
 from dtrace import DTraceConsumer
 
 SCRIPT = 'sysinfo:::readch { @bytes[execname] = sum(arg0); }'
@@ -17,7 +17,7 @@ def main():
     """
     Run DTrace...
     """
-    print 'Hint: if you don\'t get any output try running it with pfexec...'
+    print('Hint: if you don\'t get any output try running it with pfexec...')
 
     consumer = DTraceConsumer()
     consumer.run(SCRIPT, 4)

--- a/examples/read_distribution_quantize.py
+++ b/examples/read_distribution_quantize.py
@@ -7,7 +7,7 @@ Created on Mar 28, 2012
 
 @author: tmetsch
 """
-
+from __future__ import print_function
 from dtrace import DTraceConsumer
 
 # SCRIPT = 'io:::start { @bytes = quantize(args[0]->b_bcount); }'
@@ -18,10 +18,10 @@ def my_walk(action, identifier, key, values):
     """
     Walk the aggregrate.
     """
-    print key
+    print(key)
     for item in values:
         if item[0][0] > 0 and item[1] > 0:
-            print '%8s %20s' % (item[0][0], item[1])
+            print('%8s %20s' % (item[0][0], item[1]))
 
 
 def main():

--- a/examples/syscall_by_zone.py
+++ b/examples/syscall_by_zone.py
@@ -7,7 +7,7 @@ Created on Oct 10, 2011
 
 @author: tmetsch
 """
-
+from __future__ import print_function
 from dtrace import DTraceConsumerThread
 import time
 
@@ -18,7 +18,7 @@ def my_walk(action, identifier, key, value):
     """
     A simple aggregate walker.
     """
-    print 'Zone "{0:s}" made {1:d} syscalls.'.format(key[0], value)
+    print('Zone "{0:s}" made {1:d} syscalls.'.format(key[0], value))
 
 
 def main():

--- a/examples/syscall_count_own_walk.py
+++ b/examples/syscall_count_own_walk.py
@@ -8,7 +8,7 @@ Created on Oct 10, 2011
 
 @author: tmetsch
 """
-
+from __future__ import print_function
 from dtrace import DTraceConsumer
 
 SCRIPT = 'syscall:::entry { @num[execname] = count(); }'
@@ -18,7 +18,7 @@ def my_walk(action, identifier, key, value):
     """
     Aggregate walker.
     """
-    print '>', identifier, key, value
+    print('>', identifier, key, value)
 
 
 def main():

--- a/examples/syscall_lquantize.py
+++ b/examples/syscall_lquantize.py
@@ -7,7 +7,7 @@ Created on Mar 28, 2012
 
 @author: tmetsch
 """
-
+from __future__ import print_function
 from dtrace import DTraceConsumer
 
 SCRIPT = 'syscall::read:entry { @dist[execname] = lquantize(arg0, 0, 12, 2); }'
@@ -17,10 +17,10 @@ def my_walk(action, identifier, key, values):
     """
     Walk the aggregate.
     """
-    print key
+    print(key)
     for item in values:
         if item[0][0] > 0:
-            print '%8s %20s' % (item[0][0], item[1])
+            print('%8s %20s' % (item[0][0], item[1]))
 
 
 def main():

--- a/setup.py
+++ b/setup.py
@@ -10,6 +10,7 @@ Created on Oct 10, 2011
 
 from distutils.core import setup
 from distutils.extension import Extension
+import sys
 try:
     from Cython.Build import build_ext, cythonize
 
@@ -17,7 +18,7 @@ try:
     EXT_MODULES = cythonize([Extension("dtrace", ["dtrace_cython/dtrace_h.pxd",
                                                   "dtrace_cython/consumer.pyx"],
                              libraries=["dtrace"])],
-        language_level=2)
+        language_level=sys.version_info.major)
 
 except ImportError:
     BUILD_EXTENSION = {}

--- a/setup.py
+++ b/setup.py
@@ -11,12 +11,13 @@ Created on Oct 10, 2011
 from distutils.core import setup
 from distutils.extension import Extension
 try:
-    from Cython.Distutils import build_ext
+    from Cython.Build import build_ext, cythonize
 
     BUILD_EXTENSION = {'build_ext': build_ext}
-    EXT_MODULES = [Extension("dtrace", ["dtrace_cython/dtrace_h.pxd",
-                                        "dtrace_cython/consumer.pyx"],
-                             libraries=["dtrace"])]
+    EXT_MODULES = cythonize([Extension("dtrace", ["dtrace_cython/dtrace_h.pxd",
+                                                  "dtrace_cython/consumer.pyx"],
+                             libraries=["dtrace"])],
+        language_level=2)
 
 except ImportError:
     BUILD_EXTENSION = {}

--- a/tests/dtrace_ctypes_test.py
+++ b/tests/dtrace_ctypes_test.py
@@ -27,7 +27,7 @@ class TestDTraceConsumer(unittest.TestCase):
 
     def test_run_for_sanity(self):
         self.consumer.run(SCRIPT)
-        self.assertEquals(self.out, 'Hello World')
+        self.assertEqual(self.out, b'Hello World')
 
     def _get_output(self, data, arg):
         tmp = c_char_p(data.contents.dtbda_buffered).value.strip()

--- a/tests/dtrace_cython_test.py
+++ b/tests/dtrace_cython_test.py
@@ -39,7 +39,7 @@ class TestDTraceConsumer(unittest.TestCase):
 
     def test_run_for_sanity(self):
         self.consumer.run(SCRIPT)
-        self.assertEquals('Hello World', self.out)
+        self.assertEqual(b'Hello World', self.out)
 
     def _get_output(self, tmp):
         self.out = tmp

